exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinSpillControlImpl.java

```diff
@@ -18,39 +18,211 @@
+      // each partition needs at least one internal build batch, and a minimum hash-table
+      // adding "min batches" to create some safety slack
+      if ( memoryAvailableNow >
+             ( numPartitions + minBatchesInAvailableMemory ) * (batchSize + hashTableSize) ) {
+        break; // got enough memory
+      }
+    }
+    logger.debug("Spill control chosen to use {} partitions", numPartitions);
+  }
+
+  @Override
+  public long getBuildReservedMemory() {
+    return 0;
+  }
+
+  @Override
+  public long getMaxReservedMemory() {
+    return 0;
+  }
+
+  @Override
+  public String makeDebugString() {
+    return "Spill Control " + HashJoinMemoryCalculatorImpl.HashJoinSpillControl.class.getCanonicalName();
+  }
+
+  @Nullable
+  @Override
+  public HashJoinMemoryCalculator.PostBuildCalculations next() {
+    return new SpillControlPostBuildCalculationsImpl(recordsPerPartitionBatchProbe,allocator,recordsPerBatch,minBatchesInAvailableMemory,batchMemoryManager);
+  }
+
+  @Override
+  public HashJoinState getState() {
+    return HashJoinState.BUILD_SIDE_PARTITIONING;
+  }
+
+
+  /**
+   *   The purpose of this class is to provide the method {@link #shouldSpill} that ensures that enough memory is available to
+   *   hold all the probe incoming batches for those partitions that spilled (else need to spill more of them, for more memory).
+   */
+  public static class SpillControlPostBuildCalculationsImpl implements HashJoinMemoryCalculator.PostBuildCalculations {
+    private static final Logger logger = LoggerFactory.getLogger(SpillControlPostBuildCalculationsImpl.class);
+
+    private final int recordsPerPartitionBatchProbe;
+    private BufferAllocator allocator;
+    private int recordsPerBatch;
+    private int minBatchesInAvailableMemory;
+    private RecordBatchMemoryManager batchMemoryManager;
+    private int numPartitionsSpilled;
+    private boolean probeEmpty;
+
+    public SpillControlPostBuildCalculationsImpl(final int recordsPerPartitionBatchProbe,
+                                                 BufferAllocator allocator, int recordsPerBatch, int minBatchesInAvailableMemory, RecordBatchMemoryManager batchMemoryManager) {
+      this.allocator = allocator;
+      this.recordsPerBatch = recordsPerBatch;
+      this.minBatchesInAvailableMemory = minBatchesInAvailableMemory;
+      this.batchMemoryManager = batchMemoryManager;
+      this.recordsPerPartitionBatchProbe = recordsPerPartitionBatchProbe;
+    }
+
+    @Override
+    public void initialize(boolean hasProbeData, int numPartitionsSpilled) {
+      this.probeEmpty = hasProbeData;
+      this.numPartitionsSpilled = numPartitionsSpilled;
+    }
+
+    @Override
+    public int getProbeRecordsPerBatch() {
+      return recordsPerPartitionBatchProbe;
+    }
+
+    @Override
+    public boolean shouldSpill() {
+      if ( numPartitionsSpilled == 0 ) { return false; } // no extra memory is needed if all the build side is in memory
```

Comment:
This new method (used only for "duplicate skipping semi join") is simpler code than before, relying on the allocator's "getAllocatedMemory" to get the current memory state, no need to calculate in detail every hash table size, build side batches, other probe side batches, etc.

* Indeed the "incoming" batch read from the spill is not specifically accounted for. However there's a "slack" (+ min batches in memory) that can cover that small gap.
* All should be covered by `allocator.getAllocatedMemory()`, no need to reserve for that.
* This "post" method cares about each spilled partition having one in-memory internal probe batch (else spill one more). That the `numPartitionsSpilled * batchSize` that is being checked.
* Output batch: This is `batchMemoryManager.getOutputBatchSize()` .

With the new change to use the `partitionStatSet` to get the number of spilled partitions - how is this number going to be updated - in case the code returns true (i.e. spill one more) and then try again ?  Where are those stats being updated ?

Reply:
- "Indeed the incoming batch read from the spill is not specifically accounted for. However there's a "slack" (+ min batches in memory) that can cover that small gap." We should not be introducing another tuning parameter with minBatchesInMemory, we already have the safetyFactor and fragmentationFactor tuning parameters. How about we do something like the following psuedo code: 

```
if (firstCycle) { 
   partitionsMemoryUsage = allocator.getAllocatedMemory() ;
   probeReserveSpace = 0;
} else {
   partitionsMemoryUsage = allocator.getAllocatedMemory() - probeBatch.getSize();
   probeReserveSpace = probeSizePredictor.predictBatchSize(maxNumProbeRecords, false)
}
```
This would allow us to account for the probe data safely while using the existing safetyFactor and fragementationFactor parameters and would allow us to remove minBatchesInMemory.

 - PartitionStatSet collects all the stats from the HashPartitions directly. So if a HashPartition is spilled the PartitionStatSet will know. I've unit tested this and the other memory calculators use this logic without issue. The idea was to decouple the stats needed for memory calculations from the rest of the HashJoin operator, so that the memory calculators can be unit tested.